Library/Homebrew/keg_relocate.rb

```diff
@@ -44,12 +50,107 @@ def replace_text!(text)
 
       any_changed = T.let(nil, T.nilable(String))
       sorted_keys.each do |key|
-        changed = text.gsub!(key, replacements.fetch(key))
+        # Smart replacement logic for /usr/local prefix when we have keg context
+        if should_use_smart_replacement?(key, replacements[key])
+          changed = smart_replace_usr_local!(text, key, replacements.fetch(key))
+        else
+          changed = text.gsub!(key, replacements.fetch(key))
+        end
         any_changed ||= changed
       end
       !any_changed.nil?
     end
 
+    private
+
+    sig { params(pattern: T.any(String, Regexp), replacement: String).returns(T::Boolean) }
+    def should_use_smart_replacement?(pattern, replacement)
+      # Only use smart replacement when:
+      # 1. We have keg context
+      # 2. HOMEBREW_PREFIX is /usr/local  
+      # 3. We're replacing with @@HOMEBREW_PREFIX@@
+      @keg_path &&
+        HOMEBREW_PREFIX.to_s == "/usr/local" &&
+        replacement == "@@HOMEBREW_PREFIX@@" &&
+        pattern.is_a?(Regexp)
+    end
+
+    sig { params(text: String, pattern: T.any(String, Regexp), replacement: String).returns(T.nilable(String)) }
+    def smart_replace_usr_local!(text, pattern, replacement)
+      return text.gsub!(pattern, replacement) unless pattern.is_a?(Regexp)
```

Comment:
```suggestion
    sig { params(text: String, pattern: Regexp, replacement: String).returns(T.nilable(String)) }
    def smart_replace_usr_local!(text, pattern, replacement)
```
and use `should_use_smart_replacement?` to detect whether pattern is a Regexp